lib/active_encode/engine_adapters/ffmpeg_adapter.rb

```diff
@@ -0,0 +1,185 @@
+require 'fileutils'
+require 'nokogiri'
+
+module ActiveEncode
+  module EngineAdapters
+    class FfmpegAdapter
+      WORK_DIR = ENV["ENCODE_WORK_DIR"] || "encodes" # Should read from config
+
+      def create(input_url, options = {})
+        new_encode = ActiveEncode::Base.new(input_url, options)
+        new_encode.id = SecureRandom.uuid
+        new_encode.state = :running
+        new_encode.current_operations = []
+        new_encode.percent_complete = 10
+        new_encode.errors = []
+        new_encode.created_at = Time.new
+        new_encode.updated_at = Time.new
+
+        new_encode.input.url = input_url
+        new_encode.input.created_at = new_encode.created_at
+        new_encode.input.updated_at = new_encode.created_at
+
+        new_encode.output = []
+
+        # Create a working directory that holds all output files related to the encode
+        FileUtils.mkdir_p working_path("", new_encode.id)
+
+        # Extract technical metadata from input file
+        `mediainfo --Output=XML --LogFile=#{working_path("input_metadata", new_encode.id)} #{input_url}`
+        new_encode.input.assign_tech_metadata get_tech_metadata(working_path("input_metadata", new_encode.id))
+
+        # Run the ffmpeg command and save its pid
+        command = ffmpeg_command(new_encode, options)
+        pid = Process.spawn(command)
+        File.open(working_path("pid", new_encode.id), 'w') { |file| file.write pid }
+        new_encode.input.id = pid
+
+        # Prevent zombie process
+        Process.detach(pid)
+
+        new_encode
+      end
+
+      # Return encode object from file system
+      def find(id, opts={})
+        encode_class = opts[:cast]
+        encode = ActiveEncode::Base.new(nil, opts)
+        encode.id = id
+        encode.errors = []
+        encode.current_operations = []
+        encode.created_at = File.mtime working_path("pid", id)
+        encode.updated_at = File.mtime working_path("progress", id)
+
```

Comment:
This behavior is now changed, but I believe there will always be a negligible difference because we're relying on filesystem timestamp. If there's a way to minimize this difference that'd be great.